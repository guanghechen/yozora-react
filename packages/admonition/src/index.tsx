import cn from 'clsx'
import PropTypes from 'prop-types'
import React from 'react'

export interface AdmonitionProps {
  /**
   * Admonition keyword.
   */
  keyword?: string
  /**
   * Admonition title
   */
  title?: React.ReactNode
  /**
   * Admonition contents.
   */
  children?: React.ReactNode
  /**
   * Root css class of the component.
   */
  className?: string
  /**
   * Root css style.
   */
  style?: React.CSSProperties
}

/**
 * Render yozora `admonition`.
 *
 * @see https://www.npmjs.com/package/@yozora/ast#admonition
 * @see https://www.npmjs.com/package/@yozora/tokenizer-admonition
 */
export function Admonition(props: AdmonitionProps): React.ReactElement {
  const { className, style, keyword = 'default', title, children } = props

  let modifier = keyword.trim().toLowerCase()
  switch (modifier) {
    case 'note':
      modifier = 'default'
      break
    case 'important':
      modifier = 'info'
      break
    case 'tip':
      modifier = 'success'
      break
    case 'caution':
      modifier = 'warning'
      break
    case 'error':
      modifier = 'danger'
      break
  }

  return (
    <div
      className={cn(
        'yozora-admonition',
        'yozora-admonition--' + modifier,
        className,
      )}
      style={style}
    >
      <div key="heading" className="yozora-admonition__heading">
        <h5>{title}</h5>
      </div>
      <div key="body" className="yozora-admonition__body">
        {children}
      </div>
    </div>
  )
}

Admonition.propTypes = {
  children: PropTypes.node,
  className: PropTypes.string,
  keyword: PropTypes.string,
  style: PropTypes.object,
  title: PropTypes.node,
}

Admonition.displayName = 'YozoraAdmonition'
export default Admonition
