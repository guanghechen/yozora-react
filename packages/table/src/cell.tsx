import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'


/**
 * Props for table cell
 */
export interface TableCellProps
  extends React.HTMLAttributes<HTMLTableDataCellElement | HTMLTableHeaderCellElement> {
  /**
   * Content of the table cell
   */
  children: React.ReactNode
  /**
   * Whether is the table header cell
   */
  isHeader?: boolean
  /**
   * Table cell content align
   */
  align?: 'left' | 'center' | 'right'
}


const TH = styled.th`
  padding: var(--md-table-cell-padding, 0.4rem 0.8rem);
  border: 1px solid var(--md-table-border-color);
`


const TD = styled.td`
  padding: var(--md-table-cell-padding, 0.4rem 0.8rem);
  border: 1px solid var(--md-table-border-color);
`


/**
 *
 * @param props
 */
export const TableCell = React.forwardRef<
  HTMLTableDataCellElement | HTMLTableHeaderCellElement,
  TableCellProps
>(
  (props, forwardRef): React.ReactElement => {
    const { children, isHeader = false, align = 'center', ...cellProps } = props
    const T = isHeader ? TH : TD

    return (
      <T { ...cellProps } align={ align } ref={ forwardRef }>
        { children }
      </T>
    )
  }
)



TableCell.propTypes = {
  children: PropTypes.node.isRequired,
  isHeader: PropTypes.bool,
  align: PropTypes.oneOf<'left' | 'center' | 'right'>(['left', 'center', 'right']),
}


TableCell.displayName = 'TableCell'


export default TableCell
