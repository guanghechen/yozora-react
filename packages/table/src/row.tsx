import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'


/**
 * Props for table row
 */
export interface TableRowProps extends React.HTMLAttributes<HTMLTableRowElement> {
  /**
   * Table row content
   */
  children: React.ReactNode
}


const Container = styled.tr`
`


/**
 *
 * @param props
 */
export const TableRow = React.forwardRef<HTMLTableRowElement, TableRowProps>(
  (props, forwardRef): React.ReactElement => {
    return (
      <Container { ...props } ref={ forwardRef } />
    )
  }
)



TableRow.propTypes = {
  children: PropTypes.node.isRequired,
}


TableRow.displayName = 'TableRow'


export default TableRow
