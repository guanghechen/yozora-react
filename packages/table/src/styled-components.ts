import type { YozoraTableTheme } from './theme'
import 'styled-components'

declare module 'styled-components' {
  export interface DefaultTheme {
    yozora: YozoraTheme
  }

  interface YozoraTheme {
    table?: YozoraTableTheme
  }
}
