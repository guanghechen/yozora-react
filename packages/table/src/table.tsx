import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'


/**
 * Props for table
 */
export interface TableProps extends React.TableHTMLAttributes<HTMLTableElement> {
  /**
   * Table head
   */
  head: React.ReactNode
  /**
   * Table body
   */
  body?: React.ReactNode
}


const Container = styled.table`
  display: block;
  width: var(--md-table-width, max-content);
  max-width: 100%;
  overflow: var(--md-table-overflow, auto);
  margin: var(--md-table-margin, 0 0 1rem);
  border-spacing: 0;
  border-collapse: collapse;
`


const THead = styled.thead`
  box-sizing: border-box;
`


const TBody = styled.tbody`
  box-sizing: border-box;
  tr {
    background: var(--md-table-row-bg)
  }
  tr:nth-child(2n) {
    background: var(--md-table-row-even-bg)
  }
`


/**
 *
 * @param props
 */
export const Table = React.forwardRef<HTMLTableElement, TableProps>(
  (props, forwardRef): React.ReactElement => {
    const { head, body, children, ...tableProps } = props
    return (
      <Container { ...tableProps } ref={ forwardRef }>
        <THead>{ head }</THead>
        <TBody>{ body }</TBody>
      </Container>
    )
  }
)


Table.propTypes = {
  head: PropTypes.node.isRequired,
  body: PropTypes.node,
}


Table.displayName = 'Table'


export default Table
