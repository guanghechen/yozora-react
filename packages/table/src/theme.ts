import type { CSSProperties, DefaultTheme } from 'styled-components'


/**
 * yozora `table` theme
 */
export interface YozoraTableTheme {
  /**
   * Table width
   * @default 'max-content'
   */
  width?: CSSProperties['width'] | string
  /**
   * Table overflow
   * @default 'auto'
   */
  overflow?: CSSProperties['overflow'] | string
  /**
   * Table margin
   * @default '0 0 1em'
   */
  margin?: CSSProperties['margin'] | string
  /**
   * Table border spacing
   * @default 0
   */
  borderSpacing?: CSSProperties['borderSpacing'] | string
  /**
   * Table border collapse
   * @default 'collapse'
   */
  borderCollapse?: CSSProperties['borderCollapse'] | string
}


/**
 * Default yozora `table` theme
 */
export const defaultTableTheme: YozoraTableTheme = {
  width: 'max-content',
  overflow: 'auto',
  margin: '0 0 1em',
  borderSpacing: 0,
  borderCollapse: 'collapse',
}


/**
 * Get `table` style
 * @param key
 * @param defaultTheme
 */
export function getTableStyle(
  key: keyof YozoraTableTheme,
  defaultTheme: YozoraTableTheme = defaultTableTheme,
) {
  // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
  return (props: { theme: DefaultTheme }) => {
    const { yozora } = props.theme
    if (
      yozora == null ||
      yozora.table == null ||
      yozora.table[key] == null
    ) return defaultTheme[key]
    return yozora.table[key]
  }
}
