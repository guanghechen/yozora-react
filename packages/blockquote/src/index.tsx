import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import './styled-components'
import { defaultBlockquoteTheme, getBlockquoteStyle } from './theme'
export * from './theme'


/**
 * Props for creating Blockquote
 */
export interface BlockquoteProps extends React.BlockquoteHTMLAttributes<HTMLDivElement> {
  /**
   * Blockquote content
   */
  children: React.ReactNode
}


/**
 * Render Blockquote content
 *
 * @param props
 */
export const Blockquote = React.forwardRef<HTMLDivElement, BlockquoteProps>(
  (props, forwardRef): React.ReactElement => {
    return (
      <Container { ...props } ref={ forwardRef } />
    )
  }
)


Blockquote.propTypes = {
  children: PropTypes.node.isRequired,
}


Blockquote.displayName = 'YozoraBlockquote'
export default Blockquote


const Container = styled.blockquote`
  padding: ${ getBlockquoteStyle('padding') };
  border-left: 0.25em solid ${ getBlockquoteStyle('borderColor') };
  margin: ${ getBlockquoteStyle('margin') };
  background:  ${ getBlockquoteStyle('background') };
  color: ${ getBlockquoteStyle('color') };
`


Container.defaultProps = {
  theme: { yozora: { blockquote: defaultBlockquoteTheme } }
}


export const BlockquoteClasses = {
  container: `${ Container }`,
}
