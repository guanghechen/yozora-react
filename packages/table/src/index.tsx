import PropTypes from 'prop-types'
import React from 'react'
import styled from 'styled-components'
import './styled-components'
import { defaultTableTheme, getTableStyle } from './theme'

export * from './theme'

/**
 * Props for table
 */
export interface TableProps
  extends React.TableHTMLAttributes<HTMLTableElement> {
  children?: React.ReactNode[]
}

/**
 *
 * @param props
 */
export const Table = React.forwardRef<HTMLTableElement, TableProps>(
  (props, forwardRef): React.ReactElement => {
    const { children, ...htmlProps } = props
    const [headRows, ...bodyRows] = children
    return (
      <Container {...htmlProps} ref={forwardRef}>
        <thead>{headRows}</thead>
        <tbody>{bodyRows}</tbody>
      </Container>
    )
  },
)

Table.propTypes = {
  children: PropTypes.arrayOf(PropTypes.node).isRequired,
}

Table.displayName = 'YozoraTable'
export default Table

const Container = styled.table`
  display: block;
  max-width: 100%;
  width: ${getTableStyle('width')};
  overflow: ${getTableStyle('overflow')};
  margin: ${getTableStyle('margin')};
  border-spacing: ${getTableStyle('borderSpacing')};
  border-collapse: ${getTableStyle('borderCollapse')};
`

Container.defaultProps = {
  theme: { yozora: { table: defaultTableTheme } },
}

export const TableClasses = {
  container: `${Container}`,
}
