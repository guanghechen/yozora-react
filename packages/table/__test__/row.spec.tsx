import React from 'react'
import { mount, shallow } from 'enzyme'
import { TableCell, TableRow } from '../src'


describe('basic rendering case', () => {
  const errorLogger = jest
    .spyOn(global.console, 'error')
    .mockImplementation((...args) => {
      throw new Error(args.join(' '))
    })

  afterAll(() => {
    errorLogger.mockRestore()
  })

  it('render a simple content', () => {
    const text = 'Hello, world!'
    const wrapper = shallow(
      <table>
        <TableRow>
          <TableCell isHeader={ true }>
            <span>{ text }</span>
          </TableCell>
        </TableRow>
      </table>
    )
    expect(wrapper.text()).toEqual(text)
  })

  it('render with custom className', () => {
    const text = 'Hello, world!'
    const className = 'custom-list-item'
    const wrapper = shallow(
      <table>
        <TableRow>
          <TableCell isHeader={ false } className={ className }>
            <span>{ text }</span>
          </TableCell>
        </TableRow>
      </table>
    )
    expect(wrapper.find('.' + className) != null).toEqual(true)
    expect(wrapper.text()).toEqual(text)
  })

  it('children is required', () => {
    for (const value of [undefined, null] as any[]) {
      expect(() => {
        mount(
          <table>
            <TableRow>
              <TableCell isHeader={ true }>
                { value }
              </TableCell>
            </TableRow>
          </table>
        )
      }).toThrow(/Failed prop type: The prop `children` is marked as required/)
    }
  })

  it('forward ref', () => {
    const ref = React.createRef<HTMLTableRowElement>()
    const wrapper = mount(
      <table>
        <tbody>
          <TableRow ref={ ref } data-value="waw">
            <TableCell isHeader={ false }>
              1
            </TableCell>
          </TableRow>
        </tbody>
      </table>
    )

    const o = wrapper.find('tr').getDOMNode()
    expect(o).toEqual(ref.current)
    expect(o.getAttribute('data-value')).toEqual('waw')
  })

  it('snapshot', () => {
    const wrapper = mount(
      <table>
        <tbody>
          <TableRow style={ { color: 'orange', fontSize: '16px' } }>
            <TableCell isHeader={ false }>
              some text1
              <span>some text2</span>
            </TableCell>
          </TableRow>
        </tbody>
      </table>
    )
    expect(wrapper).toMatchSnapshot()
  })
})
