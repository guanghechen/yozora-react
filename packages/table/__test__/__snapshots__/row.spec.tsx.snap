// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic rendering case snapshot 1`] = `
.c0 {
  padding: var(--md-table-cell-padding,0.4rem 0.8rem);
  border: 1px solid var(--md-table-border-color);
}

<table>
  <tbody>
    <TableRow
      style={
        Object {
          "color": "orange",
          "fontSize": "16px",
        }
      }
    >
      <styled.tr
        style={
          Object {
            "color": "orange",
            "fontSize": "16px",
          }
        }
      >
        <tr
          className=""
          style={
            Object {
              "color": "orange",
              "fontSize": "16px",
            }
          }
        >
          <TableCell
            isHeader={false}
          >
            <styled.td
              align="center"
            >
              <td
                className="c0"
              >
                some text1
                <span>
                  some text2
                </span>
              </td>
            </styled.td>
          </TableCell>
        </tr>
      </styled.tr>
    </TableRow>
  </tbody>
</table>
`;
